rpc-broker/src/msg.c

```diff
@@ -0,0 +1,286 @@
+        else
+            DBUS_BROKER_EVENT("%s <%s>", req_msg, "Passed request");
+    }
+
+    return policy;
+}
+
+void debug_raw_buffer(char *buf, int rbytes)
+{
+    char tmp[DBUS_MSG_LEN] = { '\0' };
+
+    int i;
+    for (i=0; i < rbytes; i++) {
+        if (isalnum(buf[i]))
+            tmp[i] = buf[i];
+        else
+            tmp[i] = '-';
+    }
+
+    DBUS_BROKER_EVENT("5555: %s", tmp);
+}
+
+/*
+ * This is an opaque exchange reading off from the receiving end of a raw-dbus
+ * connection.
+ */
+int exchange(int rsock, int ssock, uint16_t domid, bool is_client)
+{
+    struct dbus_message dmsg;
+    int total, rbytes, len;
+    char buf[DBUS_MSG_LEN] = { 0 };
+
+    total = 0;
+    rbytes = 0;
+
+    while ((rbytes = recv(rsock, buf, DBUS_MSG_LEN, 0)) > 0) {
+        if (rbytes > DBUS_COMM_MIN && is_client) {
+
+            len = dbus_message_demarshal_bytes_needed(buf, rbytes);
+
+            if (len == rbytes) {
+                if (convert_raw_dbus(&dmsg, buf, len) < 1)
+                    return -1;
+                if (broker(&dmsg, domid) <= 0)
+                    return -1;
+            }
+#ifdef DEBUG
+        debug_raw_buffer(buf, rbytes);
+#endif
+        }
+
+        total += rbytes;
+        send(ssock, buf, rbytes, 0);
+    }
+
+    return total;
+}
+
+static inline char *get_db_vm_path(DBusConnection *conn, uint16_t domid)
+{
+#ifdef HAVE_XENSTORE
+    size_t len;
+    char path[256] = { 0 };
+    char *uuid;
+    struct xs_handle *xsh;
+
+    xsh = xs_open(XS_OPEN_READONLY);
+
+    if (!xsh)
+       return NULL;
+
+    snprintf(path, 255, "/local/domain/%d/vm", domid);
+
+    uuid = (char *) xs_read(xsh, XBT_NULL, path, &len);
+    xs_close(xsh);
+
+    return uuid;
+#endif
+    return NULL;
+}
+
+static int filter_if_bool(DBusConnection *conn, char *uuid,
+                          char *bool_cond, int bool_flag)
+{
+    char *arg, *attr_cond;
+    int rc;
+
+    rc = 0;
+    arg = NULL;
+    attr_cond = NULL;
+
+    DBUS_REQ_ARG(arg, "%s/%s", uuid, bool_cond);
+    attr_cond = db_query(conn, arg);
+    free(arg);
+
+    if (!attr_cond)
+        return -1;
+
+    if ((!strcmp("true", attr_cond) && bool_flag == 0) ||
+        (!strcmp("false", attr_cond) && bool_flag == 1))
+        rc = -1;
+
+    free(attr_cond);
+
+    return rc;
+}
+
+static int filter_domtype(DBusConnection *conn, char *uuid,
+                                                char *policy_domtype)
+{
+    int ret;
+    char *arg, *dom_type;
+
+    DBUS_REQ_ARG(arg, "%s/type", uuid);
+    dom_type = db_query(conn, arg);
+    free(arg);
+
+    ret = 0;
+    if (dom_type && strcmp(policy_domtype, dom_type))
+        ret = -1;
+
+    if (dom_type)
+        free(dom_type);
+
+    return ret;
+}
+
+/*
+ * The main policy filtering function, compares the policy-rule against the dbus
+ * request being made.
+ */
+int filter(struct rule *policy_rule, struct dbus_message *dmsg, uint16_t domid)
```

Comment:
According to the comment, that function sounds like a `policy_is_allowed` that should be located in policy.c and return a `bool`.
But looking at the code, I just don't understand what this function does...